### PR TITLE
Support remove blocks on worker for pagestore when free/delete file

### PR DESCRIPTION
**Pull request by fffanyang**
### What changes are proposed in this pull request?

Support removeBlock for pagestore.

### Why are the changes needed?

To remove metadata and data of blocks and pages on worker when free or delete a file.

### Does this PR introduce any user facing changes?

N/A

### CODE REVIEW

**Comment by maobaolong** [comment]
@dbw9580 As this is the first PR from @fffanyang , she has been submit a cla signed mail to `[cla@alluxio.org](mailto:cla@alluxio.org)`, PTAL.

**Comment by huanghua78** [comment]
> @dbw9580 As this is the first PR from @fffanyang , she has been submit a cla signed mail to `[cla@alluxio.org](mailto:cla@alluxio.org)`, PTAL.

Thanks to @fffanyang for submitting PR and contributing to Alluxio.

**Comment by fffanyang** [comment]
@maobaolong Could you help re-run the unit test, this may be a flaky test. Thank you in advance!

File: core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java [hunk 1]
```diff
@@ -507,4 +509,29 @@ private int getDirIndexOfBlock(long blockId) {
         .getDir()
         .getDirIndex();
   }
+
+  private void removeBlockInternal(long sessionId, long blockId, long timeoutMs)
+      throws IOException {
```
Comment (maobaolong) [inline core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java:514]: Should us replace the `IOException` to the concrete exception, something like `DeadlineExceededException `

**Review by maobaolong — COMMENTED** [review]
@fffanyang It Looks Good To Me overall, left a couple of comments inline and something need @dbw9580 to take another look.

File: core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java [hunk 2]
```diff
@@ -507,4 +509,29 @@ private int getDirIndexOfBlock(long blockId) {
...
+    Optional<BlockLock> optionalLock =
+        mLockManager.tryAcquireBlockLock(sessionId, blockId, BlockLockType.WRITE,
+        timeoutMs, TimeUnit.MILLISECONDS);
+    if (!optionalLock.isPresent()) {
+      throw new DeadlineExceededException("Cannot acquire lock to remove block");
+    }
+    try (BlockLock blockLock = optionalLock.get()) {
+      Set<PageId> pageIds;
+      try (LockResource metaLock = new LockResource(mPageMetaStore.getLock().readLock())) {
```
Comment (dbw9580) [inline core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java:523]: You need a write lock to delete blocks and pages.

File: core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java [hunk 3]
```diff
@@ -507,4 +509,29 @@ private int getDirIndexOfBlock(long blockId) {
...
+        if (mPageMetaStore.hasTempBlock(blockId)) {
+          throw new IllegalStateException(
```
[the remark this answers is not in the document]
Comment (dbw9580) [reply core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java:525]: This is fine. Temp blocks are removed by the `abortBlock` method, and `removeBlock` should only handle committed blocks.

File: core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java [hunk 4]
```diff
@@ -507,4 +509,29 @@ private int getDirIndexOfBlock(long blockId) {
...
+            ExceptionMessage.REMOVE_UNCOMMITTED_BLOCK.getMessage(blockId));
+        }
+        pageIds = mPageMetaStore.getBlock(blockId)
+          .orElseThrow(() -> new BlockDoesNotExistRuntimeException(blockId))
+          .getDir().getBlockPages(blockId);
+      }
+      for (PageId pageId : pageIds) {
+        mCacheManager.delete(pageId);
+      }
```
Comment (dbw9580) [inline core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java:534]: cache manager is the high level API. calling delete on cache manager will lock the metastore internally and cause a deadlock.

To delete pages of a block, you need to acquire the write lock on the page metastore, and remove them through the API of the metastore.

**Review by dbw9580 — CHANGES_REQUESTED** [review]
Thanks for the PR! some initial comments.

File: core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java (hunk 1 above)
Comment (dbw9580) [reply core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java:514]: Yes, if concrete exceptions can be specified.

File: core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java (hunk 4 above)
Comment (fffanyang) [reply core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java]: @dbw9580 please take a look again, I already fix it. Thanks a lot!

File: core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java [hunk 5]
```diff
@@ -397,15 +401,13 @@ public void moveBlock(long sessionId, long blockId, AllocateOptions moveOptions)
   public void removeBlock(long sessionId, long blockId) throws IOException {
     LOG.debug("removeBlock: sessionId={}, blockId={}", sessionId, blockId);
     // TODO(bowen): implement actual removal and replace placeholder values
```
[the remark this answers is not in the document]
Comment (fffanyang) [reply core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java:405]: done

File: core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java (hunk 1 above)
Comment (fffanyang) [reply core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java:516]: The delete func in PageStore throws IOException (pageInfo.getLocalCacheDir().getPageStore().delete(pageId);), so we still need to throw IOException here.

File: core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java (hunk 2 above)
Comment (fffanyang) [reply core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java]: @dbw9580 done

File: core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (maobaolong) [reply core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java:405]: Maybe you forgot push your commit?

File: core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java (hunk 1 above)
Comment (maobaolong) [reply core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java:516]: It seems make sense.

File: core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (fffanyang) [reply core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java]: please check again, thanks!

File: core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java [hunk 6]
```diff
@@ -507,4 +510,33 @@ private int getDirIndexOfBlock(long blockId) {
...
+          .orElseThrow(() -> new BlockDoesNotExistRuntimeException(blockId))
+          .getDir().getBlockPages(blockId);
+
+        for (PageId pageId : pageIds) {
+          PageInfo pageInfo = mPageMetaStore.removePage(pageId);
+          pageInfo.getLocalCacheDir().getPageStore().delete(pageId);
+        }
+      }
+    } catch (PageNotFoundException e) {
+      throw AlluxioRuntimeException.from(e);
```
Comment (dbw9580) [inline core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java:539]: Conversion from a `PageNotFoundException` which is a `AlluxioException` to `AlluxioRuntimeException` is not well-defined (it gets converted to a `UnknownRuntimeException`. Consider manually constructing a `NotFoundRuntimeException` from the underlying exception.

File: core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java [hunk 7]
```diff
@@ -507,4 +510,33 @@ private int getDirIndexOfBlock(long blockId) {
...
         .getDirIndex();
   }
+
+  private void removeBlockInternal(long sessionId, long blockId, long timeoutMs)
+      throws IOException {
+    Optional<BlockLock> optionalLock =
+        mLockManager.tryAcquireBlockLock(sessionId, blockId, BlockLockType.WRITE,
+        timeoutMs, TimeUnit.MILLISECONDS);
+    if (!optionalLock.isPresent()) {
+      throw new DeadlineExceededException("Cannot acquire lock to remove block");
```
Comment (dbw9580) [inline core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java:520]: Include the block ID in the exception message.

**Review by dbw9580 — CHANGES_REQUESTED** [review]
Mostly LGTM except for 2 minor issues.

Can you also please add a integration test in `PagedBlockStoreIntegrationTest` ? This can come in another PR.

File: core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java (hunk 6 above)
Comment (fffanyang) [reply core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java]: @dbw9580 Done. PTAL thanks!

File: core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java (hunk 7 above)
Comment (fffanyang) [reply core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java]: Done.

File: core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java [hunk 8]
```diff
@@ -507,4 +511,35 @@ private int getDirIndexOfBlock(long blockId) {
...
+          .orElseThrow(() -> new BlockDoesNotExistRuntimeException(blockId))
+          .getDir().getBlockPages(blockId);
+
+        for (PageId pageId : pageIds) {
+          PageInfo pageInfo = mPageMetaStore.removePage(pageId);
+          pageInfo.getLocalCacheDir().getPageStore().delete(pageId);
+        }
+      }
+    } catch (PageNotFoundException e) {
+      throw new NotFoundRuntimeException(e.getMessage(), e.getCause());
```
Comment (dbw9580) [inline core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java:542]:
```suggestion
      throw new NotFoundRuntimeException("Page not found: " + e.getMessage(), e);
```

**Review by dbw9580 — APPROVED** [review]
LGTM! Thanks for your contribution!

**Review by maobaolong — APPROVED** [review]
LGTM

**Comment by jiacheliu3** [comment]
alluxio-bot, merge this please

**Pull request merged**